crypto/aes/aesccm.c

```diff
@@ -0,0 +1,323 @@
+/**
+ * Copyright (c) 2023 Andrew Kozlik
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <string.h>
+
+#include "aesccm.h"
+#include "memzero.h"
+
+typedef struct {
+  const aes_encrypt_ctx *encrypt_ctx;
+  union {
+    // Ensure 32-bit alignment.
+    uint8_t state[AES_BLOCK_SIZE];
+    uint32_t state32[AES_BLOCK_SIZE / 4];
+  };
+  // Next position in the state where data will be added.
+  // Valid values are 0 to 15.
+  uint8_t pos;
+} cbc_mac_context;
+
+// WARNING: Caller must ensure that encrypt_ctx remains valid for the lifetime
+// of ctx.
+static void cbc_mac_init(cbc_mac_context *ctx,
+                         const aes_encrypt_ctx *encrypt_ctx) {
+  memzero(ctx, sizeof(cbc_mac_context));
+  ctx->encrypt_ctx = encrypt_ctx;
+}
+
+static AES_RETURN cbc_mac_update(cbc_mac_context *ctx, const uint8_t *data,
+                                 size_t size) {
+  if (ctx->pos != 0) {
+    while (size > 0 && ctx->pos < AES_BLOCK_SIZE) {
+      ctx->state[ctx->pos] ^= *data;
+      ctx->pos++;
+      data++;
+      size--;
+    }
+
+    if (ctx->pos != AES_BLOCK_SIZE) {
+      return EXIT_SUCCESS;
+    }
+
+    if (aes_encrypt(ctx->state, ctx->state, ctx->encrypt_ctx) != EXIT_SUCCESS) {
+      memzero(ctx, sizeof(*ctx));
+      return EXIT_FAILURE;
+    }
+    ctx->pos = 0;
+  }
+
+  size_t block_count = size >> AES_BLOCK_SIZE_P2;
+  size %= AES_BLOCK_SIZE;
+
+  if (!ALIGN_OFFSET(data, 4)) {
```

Comment:
I'm strongly against such optimizations unless we know the code is too slow for our purposes.

In short, my opinion is that the increase in speed does not worth the complexity added to the code. I have some arguments both for and against and I will write them out here, in case you don't like either of these suggestions:
  * remove the optimization completely
  * factor out `void memxor (void *destination, const void *source, size_t length)` implemented as a function or as a macro that would encapsulate the optimization ([this](https://github.com/bminor/glibc/blob/master/string/memcpy.c) or [this](https://github.com/bminor/newlib/blob/master/newlib/libc/string/memcpy.c) can server as an inspiration)

Reply:
This is just an adaptation of the existing CBC implementation in `aes_modes.c`. I left out `FAST_BUFFER_OPERATIONS`, because I don't see any portability issue, and even if there was one, other functions in trezor-crypto already rely on this feature. So this code just respects the existing convention of how AES modes are implemented.

Reply:
Ok, then.

These are the arguments for which I see:
  * As you already said, a similar code is used elsewhere in the codebase. I consider this argument to be a [fallacy](https://en.wikipedia.org/wiki/Appeal_to_tradition). The fact that there is bad code in our codebase does not justify us writing bad code as well.
  * We have no signs that the code doesn't work in terms of correctness of the results.

The arguments against which I see:
  * We have no signs that the optimized code is faster.
  * Generally I think it's better to rely on compiler to optimize a low level code instead of human.
  * The code is more difficult to understand, review and maintain because it uses "non-standard" patterns.
  * It may be my lack of knowledge of the C language but these two patterns specifically raise doubts in me:
    * ```      
        ctx->state32[3] ^= ((uint32_t *)data)[3];
        if (aes_encrypt(ctx->state, ctx->state, ctx->encrypt_ctx) != EXIT_SUCCESS) {
      ```
      At least C 1999 standard says the value of a union member, other than the last one stored, is unspecified.
    * ```ctx->state32[0] ^= ((uint32_t *)data)[0];```
      I don't know whether reinterpretation of `uint8_t*` as `uint32_t*` is defined.

As I said, I'm completely okay if the problematic patterns remain in the code but are encapsulated in a `memxor` function.

Reply:
There is no fallacy in my statement. A certain level of optimization was chosen in the existing AES code. Speculation about how that choice was reached is outside of the scope of this PR. So far no proof of anything being objectively "bad" about the existing code has been given. The new code simply respects the style and the chosen level of optimization of the existing code in order to maintain consistency. That doesn't rule out the possibility of refactoring the entire AES implementation at some point in the future if we see fit, but that is also out of scope of this PR.

Regarding the rest, see https://en.wikipedia.org/wiki/Type_punning.